src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py

```diff
@@ -337,12 +337,17 @@ def trash_incarnation_dir(self):
         except cephfs.Error as e:
             raise VolumeException(-e.args[0], e.args[1])
 
-    def remove(self, retainsnaps=False):
+    def remove(self, retainsnaps=False, internal_cleanup=False):
         if self.list_snapshots():
             if not retainsnaps:
                 raise VolumeException(-errno.ENOTEMPTY, "subvolume '{0}' has snapshots".format(self.subvolname))
         else:
             if not self.has_pending_purges:
+                if not internal_cleanup and \
+                  self.state not in [SubvolumeStates.STATE_COMPLETE, SubvolumeStates.STATE_CANCELED, SubvolumeStates.STATE_FAILED]:
```

Comment:
The STATE_FAILED and STATE_CANCELED are both handled by handle_clone_failed in the state machine which removes the entry from the index. Hence allowed the clone to be removed with force option for both.

Reply:
Done

Reply:
nit: move this to a helper?